plugin/pkg/admission/noderestriction/admission.go

```diff
@@ -230,6 +201,80 @@ func (p *Plugin) admitPod(nodeName string, a admission.Attributes) error {
+		return admission.NewForbidden(a, fmt.Errorf("unexpected type %T", a.GetObject()))
+	}
+
+	// only allow nodes to create mirror pods
+	if _, isMirrorPod := pod.Annotations[api.MirrorPodAnnotationKey]; !isMirrorPod {
+		return admission.NewForbidden(a, fmt.Errorf("pod does not have %q annotation, node %q can only create mirror pods", api.MirrorPodAnnotationKey, nodeName))
+	}
+
+	// only allow nodes to create a pod bound to itself
+	if pod.Spec.NodeName != nodeName {
+		return admission.NewForbidden(a, fmt.Errorf("node %q can only create pods with spec.nodeName set to itself", nodeName))
+	}
+	if len(pod.OwnerReferences) > 0 {
+		if len(pod.OwnerReferences) > 1 {
+			return admission.NewForbidden(a, fmt.Errorf("node %q can only create pods with a single owner reference set to itself", nodeName))
+		}
+		owner := pod.OwnerReferences[0]
+		if owner.APIVersion != v1.SchemeGroupVersion.String() ||
+			owner.Kind != "Node" ||
+			owner.Name != nodeName {
+			return admission.NewForbidden(a, fmt.Errorf("node %q can only create pods with an owner reference set to itself", nodeName))
+		}
+		if owner.Controller == nil || !*owner.Controller {
+			return admission.NewForbidden(a, fmt.Errorf("node %q can only create pods with a controller owner reference set to itself", nodeName))
+		}
+	}
+
+	// don't allow a node to create a pod that references any other API objects
+	if pod.Spec.ServiceAccountName != "" {
+		return admission.NewForbidden(a, fmt.Errorf("node %q can not create pods that reference a service account", nodeName))
+	}
+	hasSecrets := false
+	podutil.VisitPodSecretNames(pod, func(name string) (shouldContinue bool) { hasSecrets = true; return false })
+	if hasSecrets {
+		return admission.NewForbidden(a, fmt.Errorf("node %q can not create pods that reference secrets", nodeName))
+	}
+	hasConfigMaps := false
+	podutil.VisitPodConfigmapNames(pod, func(name string) (shouldContinue bool) { hasConfigMaps = true; return false })
+	if hasConfigMaps {
+		return admission.NewForbidden(a, fmt.Errorf("node %q can not create pods that reference configmaps", nodeName))
+	}
+	for _, v := range pod.Spec.Volumes {
+		if v.PersistentVolumeClaim != nil {
+			return admission.NewForbidden(a, fmt.Errorf("node %q can not create pods that reference persistentvolumeclaims", nodeName))
+		}
+	}
+
+	if p.features.Enabled(features.MirrorPodNodeRestriction) {
+		// With the feature enabled, the owner reference is required.
+		if len(pod.OwnerReferences) == 0 {
+			return admission.NewForbidden(a, fmt.Errorf("node %q can only create pods with an owner reference set to itself", nodeName))
+		}
+
+		// Mirror pod labels must be whitelisted by the namespace.
+		if len(pod.Labels) > 0 {
+			whitelist, err := p.allowedMirrorPodLabels(a)
+			if err != nil {
+				return err
+			}
+			for key := range pod.Labels {
+				if !whitelist.Has(key) {
+					return admission.NewForbidden(a, fmt.Errorf("node %q cannot label pods with %q in namespace %q", nodeName, key, a.GetNamespace()))
```

Comment:
is this something we want to be able to monitor for via metrics or admission annotations?

Reply:
Good point, that seems generally useful outside this feature. I assume you meant audit annotations, not admission annotations? Error messages should already be captured in the audit logs, but aren't very machine-readable. I suppose we could just add both a metric and an annotation? They're useful for different purposes...

In either case, it seems like it can come in a separate PR.